File: apps/mobile2/compose/src/commonMain/kotlin/co/typie/screen/more/stats/StatsCalculator.kt
package co.typie.screen.more.stats

import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import kotlin.math.roundToInt
import kotlin.time.Clock

val weekdayLabels = listOf("일", "월", "화", "수", "목", "금", "토")

data class StatsCharacterCountChange(
  val date: LocalDate,
  val additions: Int,
  val deletions: Int,
)

data class StreakData(
  val currentStreak: Int,
  val longestStreak: Int,
  val thisMonthDays: Int,
  val totalDays: Int,
  val avgCharactersPerDay: Int,
)

data class WeekdayData(
  val dayIndex: Int,
  val label: String,
  val totalAdditions: Int,
  val avgAdditions: Int,
  val count: Int,
)

data class StatsActivityDay(
  val date: LocalDate,
  val additions: Int,
  val deletions: Int,
) {
  val total: Int = additions + deletions
}

fun calculateStreakData(
  characterCountChanges: List<StatsCharacterCountChange>,
  totalCharacterCount: Int,
  today: LocalDate = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()).date,
): StreakData {
  val activeDates = characterCountChanges
    .asSequence()
    .filter { it.additions > 0 }
    .map { it.date.toEpochDays() }
    .toSet()

  var currentStreak = 0
  var checkDate = today.toEpochDays()

  if (checkDate !in activeDates) {
    checkDate -= 1
  }

  while (checkDate in activeDates) {
    currentStreak += 1
    checkDate -= 1
  }

  val sortedDates = activeDates.sorted()
  var longestStreak = 0
  var tempStreak = 0

  sortedDates.forEachIndexed { index, epochDay ->
    tempStreak = if (index == 0 || epochDay - sortedDates[index - 1] != 1L) 1 else tempStreak + 1
    if (tempStreak > longestStreak) {
      longestStreak = tempStreak
    }
  }

  val thisMonthDays = activeDates.count { epochDay ->
    val date = LocalDate.fromEpochDays(epochDay)
    date.year == today.year && date.month == today.month
  }

  val totalDays = activeDates.size
  val avgCharactersPerDay = if (totalDays > 0) {
    (totalCharacterCount.toDouble() / totalDays).roundToInt()
  } else {
    0
  }

  return StreakData(
    currentStreak = currentStreak,
    longestStreak = longestStreak,
    thisMonthDays = thisMonthDays,
    totalDays = totalDays,
    avgCharactersPerDay = avgCharactersPerDay,
  )
}

fun calculateWeekdayPattern(
  characterCountChanges: List<StatsCharacterCountChange>,
): List<WeekdayData> {
  val totals = MutableList(7) { 0 }
  val counts = MutableList(7) { 0 }

  characterCountChanges.forEach { change ->
    if (change.additions <= 0) {
      return@forEach
    }

    val dayIndex = (change.date.dayOfWeek.ordinal + 1) % 7
    totals[dayIndex] += change.additions
    counts[dayIndex] += 1
  }

  return weekdayLabels.mapIndexed { index, label ->
    WeekdayData(
      dayIndex = index,
      label = label,
      totalAdditions = totals[index],
      avgAdditions = if (counts[index] > 0) {
        (totals[index].toDouble() / counts[index]).roundToInt()
      } else {
        0
      },
      count = counts[index],
    )
  }
}

fun generateActivityChartDays(
  characterCountChanges: List<StatsCharacterCountChange>,
  endDate: LocalDate = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()).date,
  dayCount: Int = 90,
): List<StatsActivityDay> {
  if (dayCount <= 0) {
    return emptyList()
  }

  val startEpochDay = endDate.toEpochDays() - dayCount + 1
  val changesByDate = characterCountChanges.associateBy { it.date.toEpochDays() }

  return (startEpochDay..endDate.toEpochDays()).map { epochDay ->
    val change = changesByDate[epochDay]
    StatsActivityDay(
      date = LocalDate.fromEpochDays(epochDay),
      additions = change?.additions ?: 0,
      deletions = kotlin.math.abs(change?.deletions ?: 0),
    )
  }
}

File: legacy/mobile/lib/screens/stats/stats_calculator.dart
import 'package:jiffy/jiffy.dart';

const weekdayLabels = ['일', '월', '화', '수', '목', '금', '토'];

class StatsCharacterCountChange {
  const StatsCharacterCountChange({required this.date, required this.additions, required this.deletions});

  final Jiffy date;
  final int additions;
  final int deletions;
}

class StreakData {
  const StreakData({
    required this.currentStreak,
    required this.longestStreak,
    required this.thisMonthDays,
    required this.totalDays,
    required this.avgCharactersPerDay,
  });

  final int currentStreak;
  final int longestStreak;
  final int thisMonthDays;
  final int totalDays;
  final int avgCharactersPerDay;
}

class WeekdayData {
  const WeekdayData({
    required this.dayIndex,
    required this.label,
    required this.totalAdditions,
    required this.avgAdditions,
    required this.count,
  });

  final int dayIndex;
  final String label;
  final int totalAdditions;
  final int avgAdditions;
  final int count;
}

StreakData calculateStreakData(
  List<StatsCharacterCountChange> characterCountChanges,
  int totalCharacterCount, {
  Jiffy? now,
}) {
  final today = (now ?? Jiffy.now()).clone();
  final activeDates = {
    for (final change in characterCountChanges)
      if (change.additions > 0) _dateKey(change.date.toLocal()),
  };

  var currentStreak = 0;
  var checkDate = today;

  if (!activeDates.contains(_dateKey(today))) {
    checkDate = checkDate.subtract(days: 1);
  }

  while (activeDates.contains(_dateKey(checkDate))) {
    currentStreak++;
    checkDate = checkDate.subtract(days: 1);
  }

  var longestStreak = 0;
  var tempStreak = 0;
  final sortedDates = activeDates.toList()..sort();

  for (var i = 0; i < sortedDates.length; i++) {
    if (i == 0) {
      tempStreak = 1;
    } else {
      final prevDate = Jiffy.parse(sortedDates[i - 1]);
      final currDate = Jiffy.parse(sortedDates[i]);
      if (currDate.diff(prevDate, unit: Unit.day) == 1) {
        tempStreak++;
      } else {
        tempStreak = 1;
      }
    }
    if (tempStreak > longestStreak) {
      longestStreak = tempStreak;
    }
  }

  var thisMonthDays = 0;
  for (final date in activeDates) {
    final parsed = Jiffy.parse(date);
    if (parsed.year == today.year && parsed.month == today.month) {
      thisMonthDays++;
    }
  }

  final totalDays = activeDates.length;
  final avgCharactersPerDay = totalDays > 0 ? (totalCharacterCount / totalDays).round() : 0;

  return StreakData(
    currentStreak: currentStreak,
    longestStreak: longestStreak,
    thisMonthDays: thisMonthDays,
    totalDays: totalDays,
    avgCharactersPerDay: avgCharactersPerDay,
  );
}

List<WeekdayData> calculateWeekdayPattern(List<StatsCharacterCountChange> characterCountChanges) {
  final weekdayStats = List.generate(7, (index) => _WeekdayAccumulator(dayIndex: index, label: weekdayLabels[index]));

  for (final change in characterCountChanges) {
    if (change.additions <= 0) {
      continue;
    }

    final dayOfWeek = _toDayIndex(change.date.toLocal());
    weekdayStats[dayOfWeek]
      ..totalAdditions += change.additions
      ..count += 1;
  }

  return [
    for (final stat in weekdayStats)
      WeekdayData(
        dayIndex: stat.dayIndex,
        label: stat.label,
        totalAdditions: stat.totalAdditions,
        avgAdditions: stat.count > 0 ? (stat.totalAdditions / stat.count).round() : 0,
        count: stat.count,
      ),
  ];
}

String _dateKey(Jiffy date) => date.format(pattern: 'yyyy-MM-dd');

int _toDayIndex(Jiffy date) {
  final weekday = DateTime(date.year, date.month, date.date).weekday;
  return weekday % 7;
}

class _WeekdayAccumulator {
  _WeekdayAccumulator({required this.dayIndex, required this.label});

  final int dayIndex;
  final String label;
  int totalAdditions = 0;
  int count = 0;
}
